code/frameworks/svelte-vite/src/plugins/ts2doc.ts

```diff
@@ -0,0 +1,209 @@
+import ts from 'typescript';
+import { svelte2tsx } from 'svelte2tsx';
+import { VERSION } from 'svelte/compiler';
+import { type SvelteDataItem } from 'sveltedoc-parser';
+
+function getInitializerValue(initializer?: ts.Node) {
+  if (initializer === undefined) {
+    return undefined;
+  }
+  if (ts.isNumericLiteral(initializer)) {
+    return Number(initializer.text);
+  }
+  if (ts.isStringLiteral(initializer)) {
+    return `"${initializer.text}"`;
+  }
+  if (initializer.kind === ts.SyntaxKind.TrueKeyword) {
+    return true;
+  }
+  if (initializer.kind === ts.SyntaxKind.FalseKeyword) {
+    return false;
+  }
+  return initializer.getText();
+}
+
+/*
+ * Make sveltedoc-parser compatible data from .svelte file with svelte2tsx and TypeScript.
+ */
+export function ts2doc(fileContent: string) {
+  const shimFilename = require.resolve('svelte2tsx/svelte-shims-v4.d.ts');
+  const shimContent = ts.sys.readFile(shimFilename) || '';
+  const shimSourceFile = ts.createSourceFile(
+    shimFilename,
+    shimContent,
+    ts.ScriptTarget.Latest,
+    true
+  );
+
+  const tsx = svelte2tsx(fileContent, {
+    version: VERSION,
+    isTsFile: true,
+    mode: 'dts',
+  });
+  const currentSourceFile = ts.createSourceFile('tmp.ts', tsx.code, ts.ScriptTarget.Latest, true);
+
+  const host = ts.createCompilerHost({});
+  host.getSourceFile = (fileName, languageVersion, onError) => {
+    if (fileName === 'tmp.ts') {
+      return currentSourceFile;
+    } else if (fileName === shimContent) {
+      return shimSourceFile;
+    } else {
+      // ignore other files
+      return;
+    }
+  };
+
+  // Create a program with the custom compiler host
+  const program = ts.createProgram([currentSourceFile.fileName, shimSourceFile.fileName], {}, host);
+  const checker = program.getTypeChecker();
+
+  const propMap: Map<string, SvelteDataItem> = new Map();
+
+  const renderFunction = currentSourceFile.statements.find((statement) => {
+    return ts.isFunctionDeclaration(statement) && statement.name?.text === 'render';
+  }) as ts.FunctionDeclaration | undefined;
+  if (renderFunction === undefined) {
+    return {
+      runeUsed: false,
+      data: [],
+    };
+  }
```

Comment:
Is there a reason we're skipping here, or can we just continue even when `render` is not found?

Reply:
It's just for robustness. For instance, in case `svelte2tsx` doesn't generate a render function. This might never happen, but it's a precaution.